## 1. Data on tags over time
<p>How can we tell what programming languages and technologies are used by the most people? How about what languages are growing and which are shrinking, so that we can tell which are most worth investing time in?</p>
<p>One excellent source of data is <a href="https://stackoverflow.com/">Stack Overflow</a>, a programming question and answer site with more than 16 million questions on programming topics. By measuring the number of questions about each technology, we can get an approximate sense of how many people are using it. We're going to use open data from the <a href="https://data.stackexchange.com/">Stack Exchange Data Explorer</a> to examine the relative popularity of languages like R, Python, Java and Javascript have changed over time.</p>
<p>Each Stack Overflow question has a <strong>tag</strong>, which marks a question to describe its topic or technology. For instance, there's a tag for languages like <a href="https://stackoverflow.com/tags/r">R</a> or <a href="https://stackoverflow.com/tags/python">Python</a>, and for packages like <a href="https://stackoverflow.com/questions/tagged/ggplot2">ggplot2</a> or <a href="https://stackoverflow.com/questions/tagged/pandas">pandas</a>.</p>
<p><img src="https://assets.datacamp.com/production/project_435/img/tags.png" alt="Stack Overflow tags"></p>
<p>We'll be working with a dataset with one observation for each tag in each year. The dataset includes both the number of questions asked in that tag in that year, and the total number of questions asked in that year.</p>

In [1]:
# Load libraries
library(readr)
library(dplyr)

# Load dataset
by_tag_year <- read_csv("datasets/by_tag_year.csv")

# Inspect the dataset
by_tag_year


Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union


Rows: 40518 Columns: 4
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (1): tag
dbl (3): year, number, year_total

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


year tag                  number year_total
1     2008 .htaccess              54   58390     
2     2008 .net                 5910   58390     
3     2008 .net-2.0              289   58390     
4     2008 .net-3.5              319   58390     
5     2008 .net-4.0                6   58390     
6     2008 .net-assembly           3   58390     
7     2008 .net-core               1   58390     
8     2008 2d                     42   58390     
9     2008 32-bit                 19   58390     
10    2008 32bit-64bit             4   58390     
11    2008 3d                     73   58390     
12    2008 64bit                 149   58390     
13    2008 abap                   10   58390     
14    2008 absolute                1   58390     
15    2008 abstract                5   58390     
16    2008 abstract-class         27   58390     
17    2008 abstract-syntax-tree    6   58390     
18    2008 accelerometer           3   58390     
19    2008 access                  1   58390     
20    2008 access-control         12   58390     
21    2008 accessibility          26   58390     
22    2008 access-vba             50   58390     
23    2008 access-violation        4   58390     
24    2008 accordion               9   58390     
25    2008 acl                    11   58390     
26    2008 acrobat                10   58390     
27    2008 action                 10   58390     
28    2008 actionlistener          4   58390     
29    2008 actionmailer            3   58390     
30    2008 actionscript          136   58390     
⋮     ⋮    ⋮                    ⋮      ⋮         
40489 2018 yaml                  648   1085170   
40490 2018 yarn                  357   1085170   
40491 2018 yeoman                 36   1085170   
40492 2018 yesod                  41   1085170   
40493 2018 yield                  69   1085170   
40494 2018 yii                   269   1085170   
40495 2018 yii2                 1181   1085170   
40496 2018 yii2-advanced-app     209   1085170   
40497 2018 yocto                 288   1085170   
40498 2018 youtube               676   1085170   
40499 2018 youtube-api           473   1085170   
40500 2018 youtube-api-v3        223   1085170   
40501 2018 youtube-data-api      203   1085170   
40502 2018 yui                     5   1085170   
40503 2018 yum                    98   1085170   
40504 2018 z3                    124   1085170   
40505 2018 zend-db                11   1085170   
40506 2018 zend-form              13   1085170   
40507 2018 zend-framework        188   1085170   
40508 2018 zend-framework2       108   1085170   
40509 2018 zeromq                168   1085170   
40510 2018 z-index               107   1085170   
40511 2018 zip                   410   1085170   
40512 2018 zipfile               115   1085170   
40513 2018 zk                     35   1085170   
40514 2018 zlib                   89   1085170   
40515 2018 zoom                  196   1085170   
40516 2018 zsh                   175   1085170   
40517 2018 zurb-foundation       182   1085170   
40518 2018 zxing                  95   1085170

## 2. Now in fraction format
<p>This data has one observation for each pair of a tag and a year, showing the number of questions asked in that tag in that year and the total number of questions asked in that year. For instance, there were 54 questions asked about the <code>.htaccess</code> tag in 2008, out of a total of 58390 questions in that year.</p>
<p>Rather than just the counts, we're probably interested in a percentage: the fraction of questions that year that have that tag. So let's add that to the table.</p>

In [2]:
# Add fraction column
by_tag_year_fraction <- by_tag_year %>% mutate(fraction = number/year_total)

# Print the new table
by_tag_year_fraction

year tag                  number year_total fraction    
1     2008 .htaccess              54   58390      9.248159e-04
2     2008 .net                 5910   58390      1.012160e-01
3     2008 .net-2.0              289   58390      4.949478e-03
4     2008 .net-3.5              319   58390      5.463264e-03
5     2008 .net-4.0                6   58390      1.027573e-04
6     2008 .net-assembly           3   58390      5.137866e-05
7     2008 .net-core               1   58390      1.712622e-05
8     2008 2d                     42   58390      7.193013e-04
9     2008 32-bit                 19   58390      3.253982e-04
10    2008 32bit-64bit             4   58390      6.850488e-05
11    2008 3d                     73   58390      1.250214e-03
12    2008 64bit                 149   58390      2.551807e-03
13    2008 abap                   10   58390      1.712622e-04
14    2008 absolute                1   58390      1.712622e-05
15    2008 abstract                5   58390      8.563110e-05
16    2008 abstract-class         27   58390      4.624079e-04
17    2008 abstract-syntax-tree    6   58390      1.027573e-04
18    2008 accelerometer           3   58390      5.137866e-05
19    2008 access                  1   58390      1.712622e-05
20    2008 access-control         12   58390      2.055146e-04
21    2008 accessibility          26   58390      4.452817e-04
22    2008 access-vba             50   58390      8.563110e-04
23    2008 access-violation        4   58390      6.850488e-05
24    2008 accordion               9   58390      1.541360e-04
25    2008 acl                    11   58390      1.883884e-04
26    2008 acrobat                10   58390      1.712622e-04
27    2008 action                 10   58390      1.712622e-04
28    2008 actionlistener          4   58390      6.850488e-05
29    2008 actionmailer            3   58390      5.137866e-05
30    2008 actionscript          136   58390      2.329166e-03
⋮     ⋮    ⋮                    ⋮      ⋮          ⋮           
40489 2018 yaml                  648   1085170    5.971415e-04
40490 2018 yarn                  357   1085170    3.289807e-04
40491 2018 yeoman                 36   1085170    3.317453e-05
40492 2018 yesod                  41   1085170    3.778210e-05
40493 2018 yield                  69   1085170    6.358451e-05
40494 2018 yii                   269   1085170    2.478874e-04
40495 2018 yii2                 1181   1085170    1.088309e-03
40496 2018 yii2-advanced-app     209   1085170    1.925966e-04
40497 2018 yocto                 288   1085170    2.653962e-04
40498 2018 youtube               676   1085170    6.229439e-04
40499 2018 youtube-api           473   1085170    4.358764e-04
40500 2018 youtube-api-v3        223   1085170    2.054978e-04
40501 2018 youtube-data-api      203   1085170    1.870675e-04
40502 2018 yui                     5   1085170    4.607573e-06
40503 2018 yum                    98   1085170    9.030843e-05
40504 2018 z3                    124   1085170    1.142678e-04
40505 2018 zend-db                11   1085170    1.013666e-05
40506 2018 zend-form              13   1085170    1.197969e-05
40507 2018 zend-framework        188   1085170    1.732447e-04
40508 2018 zend-framework2       108   1085170    9.952358e-05
40509 2018 zeromq                168   1085170    1.548145e-04
40510 2018 z-index               107   1085170    9.860206e-05
40511 2018 zip                   410   1085170    3.778210e-04
40512 2018 zipfile               115   1085170    1.059742e-04
40513 2018 zk                     35   1085170    3.225301e-05
40514 2018 zlib                   89   1085170    8.201480e-05
40515 2018 zoom                  196   1085170    1.806169e-04
40516 2018 zsh                   175   1085170    1.612651e-04
40517 2018 zurb-foundation       182   1085170    1.677157e-04
40518 2018 zxing                  95   1085170    8.754389e-05

## 3. Has R been growing or shrinking?
<p>So far we've been learning and using the R programming language. Wouldn't we like to be sure it's a good investment for the future? Has it been keeping pace with other languages, or have people been switching out of it?</p>
<p>Let's look at whether the fraction of Stack Overflow questions that are about R has been increasing or decreasing over time.</p>

In [0]:
# Filter for R tags
r_over_time <- ....

# Print the new table
# .... YOUR CODE FOR TASK 3 ....

## 4. Visualizing change over time
<p>Rather than looking at the results in a table, we often want to create a visualization. Change over time is usually visualized with a line plot.</p>

In [0]:
# Load ggplot2
# .... YOUR CODE FOR TASK 4 ....

# Create a line plot of fraction over time
# .... YOUR CODE FOR TASK 4 ....

## 5. How about dplyr and ggplot2?
<p>Based on that graph, it looks like R has been growing pretty fast in the last decade. Good thing we're practicing it now!</p>
<p>Besides R, two other interesting tags are dplyr and ggplot2, which we've already used in this analysis. They both also have Stack Overflow tags!</p>
<p>Instead of just looking at R, let's look at all three tags and their change over time. Are each of those tags increasing as a fraction of overall questions? Are any of them decreasing?</p>

In [0]:
# A vector of selected tags
selected_tags <- ....

# Filter for those tags
selected_tags_over_time <- ....

# Plot tags over time on a line plot using color to represent tag
# .... YOUR CODE FOR TASK 5 ....

## 6. What are the most asked-about tags?
<p>It's sure been fun to visualize and compare tags over time. The dplyr and ggplot2 tags may not have as many questions as R, but we can tell they're both growing quickly as well.</p>
<p>We might like to know which tags have the most questions <em>overall</em>, not just within a particular year. Right now, we have several rows for every tag, but we'll be combining them into one. That means we want <code>group_by()</code> and <code>summarize()</code>.</p>
<p>Let's look at tags that have the most questions in history.</p>

In [0]:
# Find total number of questions for each tag
sorted_tags <- by_tag_year %>%
# .... YOUR CODE FOR TASK 6 ....

# Print the new table
# .... YOUR CODE FOR TASK 6

## 7. How have large programming languages changed over time?
<p>We've looked at selected tags like R, ggplot2, and dplyr, and seen that they're each growing. What tags might be <em>shrinking</em>? A good place to start is to plot the tags that we just saw that were the most-asked about of all time, including JavaScript, Java and C#.</p>

In [0]:
# Get the six largest tags
highest_tags <- head(sorted_tags$tag)

# Filter for the six largest tags
by_tag_subset <- ....

# Plot tags over time on a line plot using color to represent tag
# .... YOUR CODE FOR TASK 7 ....

## 8. Some more tags!
<p>Wow, based on that graph we've seen a lot of changes in what programming languages are most asked about. C# gets fewer questions than it used to, and Python has grown quite impressively.</p>
<p>This Stack Overflow data is incredibly versatile. We can analyze <em>any</em> programming language, web framework, or tool where we'd like to see their change over time. Combined with the reproducibility of R and its libraries, we have ourselves a powerful method of uncovering insights about technology.</p>
<p>To demonstrate its versatility, let's check out how three big mobile operating systems (Android, iOS, and Windows Phone) have compared in popularity over time. But remember: this code can be modified simply by changing the tag names!</p>

In [0]:
# Get tags of interest
my_tags <- ....

# Filter for those tags
by_tag_subset <- ....

# Plot tags over time on a line plot using color to represent tag
# .... YOUR CODE FOR TASK 8 ....